sdk/nodejs/provider/server.ts

```diff
@@ -308,6 +308,10 @@ class Server implements grpc.UntypedServiceImplementation {
             };
 
             const result = await this.provider.construct(name, type, inputs, opts);
+
+            // Wait for RPC operations to complete and disconnect.
+            await runtime.disconnect();
```

Comment:
What is the mapping of provider instance to construct request? One-to-one or one-to-many? 

disconnecting the runtime causes the rpc clients for the engine and resmon to disconnect, but I believe those connections can be re-established the next time a mon/engine client is accessed in the node runtime.

Reply:
> What is the mapping of provider instance to construct request? One-to-one or one-to-many?

Pretty sure one-to-many. The provider process will run and can receive multiple `construct` calls. We reset the values on each invocation of `construct`:

https://github.com/pulumi/pulumi/blob/ec220de5f91e0ce0eda19a5b05cb5b06350cb07d/sdk/nodejs/provider/server.ts#L250-L275

We should be able to disconnect from the monitor and engine as the next call to `construct` will re-establish connections to these.

Though, this is changing globals... Seems like we'll run into problems if there are ever _concurrent_ calls to `construct`. @pgavlin, do you know off hand if concurrent calls are possible? Should we be serializing calls to `construct` in the Node.js implementation so that only one call to `construct` runs at a time?

Reply:
> Seems like we'll run into problems if there are ever concurrent calls to construct, though not sure offhand if that's even possible given the event loop. @pgavlin, do you know off hand if concurrent calls are possible?

Concurrent calls are absolutely possible, yes. Serializing `construct` calls is a good short-term workaround, but I'm not sure that it's sustainable long-term due to the limits it places on parallelism. We might want to scope the engine connection somehow (or at least the in-flight RPCs). IIRC @lukehoban suggested investigating node's [`createContext` API](https://nodejs.org/api/vm.html#vm_vm_createcontext_contextobject_options) to see if we could run each invocation of `Construct` in its own context.

I think that we will also need to await inside of the provider's implementation of `Cancel`.